app/controllers/api/v1/apps/credentials_controller.rb

```diff
@@ -1,9 +1,11 @@
 # frozen_string_literal: true
 
 class Api::V1::Apps::CredentialsController < Api::BaseController
-  before_action -> { doorkeeper_authorize! :read }
-
   def show
-    render json: doorkeeper_token.application, serializer: REST::ApplicationSerializer, fields: %i(name website vapid_key)
+    if valid_doorkeeper_token?
+      render json: doorkeeper_token.application, serializer: REST::ApplicationSerializer, fields: %i(name website vapid_key client_id scopes)
+    else
+      render json: { error: 'The access token is invalid' }, status: 401
```

Comment:
What about using `doorkeeper_render_error`?

Reply:
I can change to that too. Here we're also doing introspection of the application, *not* the token.

How would you feel about vendoring in the [TokenInfoController](https://github.com/doorkeeper-gem/doorkeeper/blob/main/app/controllers/doorkeeper/token_info_controller.rb)? We can't use it directly because we don't expire access tokens (we probably should, but breaking change), and the doorkeeper code for testing if a token is "accessible?" is broken, but this would give us proper OAuth 2.0 token introspection as an endpoint.

Reply:
Sorry, I'm not sure how any of that relates to my remark, I was just suggesting using `doorkeeper_render_error` to render the error for consistency with everything else; it wouldn't affect the test or the success path.
```suggestion
      doorkeeper_render_error
```

As for doing actual token introspection, why not, but that sounds like it should be a different PR

Reply:
Yeah, so, this PR originally came from needing to introspect the token (for now I don't think our UI allows you to opt-out of certain permissions requested by an application

Reply:
Alright. If you think doing actual token introspection would be preferable to this PR and cover this PR entirely, I'm all for it. Otherwise, this PR does look fine now!

Reply:
I think being able to introspect the app is important (e.g., an old install & needing to recreate the app to add more permissions), but we also need working token introspection, so we could support a flow where a user decides not to give specific permissions but gives others.